File: Common/Serialization/Formatter.vb
Namespace Common

    Public MustInherit Class Formatter

        ' ToDo Should there be a machanism for fallback? Two suggestions:
        ' 1. Put a way Get and Set of Serializer can tell that they cannot serialize the object.
        ' 2. Put back CanSerializeObject so that we can check it before calling Get or Set.
        ' Current design was chosen as it is simpler, and adds more restrictions (removes the unneccessary freedom (~Vocab)).

        ''' <summary>
        ''' This method has to be called at the end of every constructor of every derived class.
        ''' </summary>
        Protected Sub Initialize()
            Me.Serializers.LockCurrentElements()
        End Sub

        Protected MustOverride Sub OnGetStarted()
        Protected MustOverride Sub OnGetFinished()
        Protected MustOverride Sub OnGetEnterContext(ByVal Name As String)
        Protected MustOverride Sub OnGetExitContext(ByVal Name As String)

        Protected MustOverride Sub OnSetStarted()
        Protected MustOverride Sub OnSetFinished()
        Protected MustOverride Sub OnSetEnterContext(ByVal Name As String)
        Protected MustOverride Sub OnSetExitContext(ByVal Name As String)

        Private Function GetSerializer(Of T)(ByVal Obj As CNullable(Of T), ByVal IsGeneric As Boolean) As Serializer
            Dim Type = GetType(T)
            If Not IsGeneric And Obj.HasValue Then
                Type = Obj.Value.GetType()
            End If

            For I As Integer = Me.Serializers.Count - 1 To 0 Step -1
                Dim S = Me.Serializers.ItemAt(I)

                If Not S.CanSerializeType(Type) Then
                    Continue For
                End If

                ' We will also return the non-generic serializers for generic serialization. The users have to check whether the returned serializer is generic or not.

                Return S
            Next
            Return Nothing
        End Function

        Private Sub SetImpl(Of T)(ByVal Name As String, ByVal Obj As T, ByVal IsGeneric As Boolean)
            Dim WasRunning = Me.IsRunning
            Dim IsRefType = Not GetType(T).IsValueType And GetType(T) <> GetType(String)

            If Not WasRunning Then
                Me.OnSetStarted()

                Me.IsRunning = True
                Assert.True(Me.GetCache.Count = 0)
                Assert.True(Me.SetCache.Count = 0)
                Assert.True(Me.ObjectsCount = 0)
            End If

            Me.OnSetEnterContext(Name)

            Try
                Dim Id = Me.ObjectsCount
                If IsRefType Then
                    Dim Tmp As (Integer, Boolean) = Nothing
                    Dim WasInCache = Me.SetCache.TryGetValue(Obj, Tmp)

                    Dim IsDone = Tmp.Item2
                    If WasInCache Then
                        Id = Tmp.Item1
                    End If

                    ' We allow more than one serialization iterations on the same object. See the serializer for Object for a use case.

                    Me.Set(NameOf(Id), Id)

                    If IsDone Then
                        Exit Sub
                    Else
                        If Not WasInCache Then
                            Me.SetCache.Add(Obj, (Id, False))
                            Me.ObjectsCount += 1
                        End If
                    End If
                End If

                Dim S = Me.GetSerializer(Of T)(Obj, IsGeneric)
                If IsGeneric Then
                    Dim ST = TryCast(S, Serializer(Of T))
                    If ST IsNot Nothing Then
                        ST.SetT(Me.SetProxy, Obj)
                    Else
                        S.Set(Me.SetProxy, Obj)
                    End If
                Else
                    Me.Set(NameOf(Serializer), S)
                    S.Set(Me.SetProxy, Obj)
                End If

                If IsRefType Then
                    Assert.True(Me.SetCache.ContainsKey(Obj))
                    Me.SetCache.Item(Obj) = (Id, True)
                End If
            Finally
                Me.OnSetExitContext(Name)

                If Not WasRunning Then
                    Me.OnSetFinished()

                    Me.IsRunning = False
                    Me.SetCache.Clear()
                    Me.ObjectsCount = 0
                End If
            End Try
        End Sub

        Private Function GetImpl(Of T)(ByVal Name As String, ByVal GObj As CNullable(Of T), ByVal IsGeneric As Boolean) As T
            Dim WasRunning = Me.IsRunning
            Dim IsRefType = Not GetType(T).IsValueType

            ' We allow T to be a value type. See the serializer for SerializationArrayChunk for a use case.

            If Not WasRunning Then
                Me.OnGetStarted()

                Me.IsRunning = True
                Assert.True(Me.GetCache.Count = 0)
                Assert.True(Me.SetCache.Count = 0)
                Assert.True(Me.ObjectsCount = 0)
            End If

            Me.OnGetEnterContext(Name)

            Try
                Dim Id = 0
                If IsRefType Then
                    Id = Me.Get(Of Integer)(NameOf(Id))

                    Dim Obj As Object = Nothing
                    If Me.GetCache.TryGetValue(Id, Obj) AndAlso Obj IsNot Nothing Then
                        Return DirectCast(Obj, T)
                    End If

                    Me.GetCache.Add(Id, Nothing)
                End If

                Dim S As Serializer
                If IsGeneric Then
                    S = Me.GetSerializer(Of T)(Nothing, True)
                Else
                    S = Me.Get(Of Serializer)(NameOf(Serializer))
                End If

                Dim ST = TryCast(S, Serializer(Of T))
                Dim R As T = Nothing

                If IsGeneric And ST IsNot Nothing Then
                    If GObj.HasValue Then
                        ST.GetT(Me.GetProxy, GObj.Value)
                        R = GObj.Value
                    Else
                        R = ST.GetT(Me.GetProxy)
                    End If
                Else
                    If GObj.HasValue Then
                        S.Get(Me.GetProxy, GObj.Value)
                        R = GObj.Value
                    Else
                        R = DirectCast(S.Get(Me.GetProxy), T)
                    End If
                End If

                If IsRefType Then
                    Dim Obj = Me.GetCache.Item(Id)
                    If Obj IsNot Nothing Then
                        Verify.True(Obj Is DirectCast(R, Object), "Two different objects returned for the same id.")
                    Else
                        Me.GetCache.Item(Id) = R
                    End If
                End If

                Return R
            Finally
                Me.OnGetExitContext(Name)

                If Not WasRunning Then
                    Me.OnGetFinished()

                    Me.IsRunning = False
                    Me.GetCache.Clear()
                    Me.ObjectsCount = 0
                End If
            End Try
        End Function

        Protected Friend Sub [Set](Of T)(ByVal Name As String, ByVal Obj As T)
            Me.SetImpl(Name, Obj, True)
        End Sub

        Protected Friend Sub [Set](ByVal Name As String, ByVal Obj As Object)
            Me.SetImpl(Name, Obj, False)
        End Sub

        Protected Friend Function [Get](Of T)(ByVal Name As String) As T
            Return Me.GetImpl(Of T)(Name, Nothing, True)
        End Function

        Protected Friend Function [Get](ByVal Name As String) As Object
            Return Me.GetImpl(Of Object)(Name, Nothing, False)
        End Function

        Protected Friend Sub [Get](Of T)(ByVal Name As String, ByVal Obj As T)
            Me.GetImpl(Of T)(Name, Obj, True)
        End Sub

        Protected Friend Sub [Get](ByVal Name As String, ByVal Obj As Object)
            Me.GetImpl(Of Object)(Name, New CNullable(Of Object)(Obj), False)
        End Sub

#Region "Serializers Property"
        Private ReadOnly _Serializers As SerializerCollection = New SerializerCollection()

        Public ReadOnly Property Serializers As SerializerCollection
            Get
                Return Me._Serializers
            End Get
        End Property
#End Region

        Private ReadOnly SetProxy As FormatterSetProxy = New FormatterSetProxy(Me)
        Private ReadOnly GetProxy As FormatterGetProxy = New FormatterGetProxy(Me)

        Private ReadOnly SetCache As Dictionary(Of Object, (Integer, Boolean)) = New Dictionary(Of Object, (Integer, Boolean))(ReferenceEqualityComparer(Of Object).Instance)
        Private ReadOnly GetCache As Dictionary(Of Integer, Object) = New Dictionary(Of Integer, Object)()
        Private ObjectsCount As Integer
        Private IsRunning As Boolean

    End Class

End Namespace

File: Common/Serialization/Formatter.cs
using System.Collections.Generic;

namespace Ks.Common
{
    public abstract partial class Formatter
    {

        // ToDo Should there be a machanism for fallback? Two suggestions:
        // 1. Put a way Get and Set of Serializer can tell that they cannot serialize the object.
        // 2. Put back CanSerializeObject so that we can check it before calling Get or Set.
        // Current design was chosen as it is simpler, and adds more restrictions (removes the unneccessary freedom (~Vocab)).

        /// <summary>
        /// This method has to be called at the end of every constructor of every derived class.
        /// </summary>
        protected void Initialize()
        {
            this.Serializers.LockCurrentElements();
        }

        protected abstract void OnGetStarted();
        protected abstract void OnGetFinished();
        protected abstract void OnGetEnterContext(string Name);
        protected abstract void OnGetExitContext(string Name);

        protected abstract void OnSetStarted();
        protected abstract void OnSetFinished();
        protected abstract void OnSetEnterContext(string Name);
        protected abstract void OnSetExitContext(string Name);

        private Serializer GetSerializer<T>(CNullable<T> Obj, bool IsGeneric)
        {
            var Type = typeof(T);
            if (!IsGeneric & Obj.HasValue)
            {
                Type = Obj.Value.GetType();
            }

            for (var I = this.Serializers.Count - 1; I >= 0; I--)
            {
                var S = this.Serializers[I];

                if (!S.CanSerializeType(Type))
                {
                    continue;
                }

                // We will also return the non-generic serializers for generic serialization. The users have to check whether the returned serializer is generic or not.

                return S;
            }
            return null;
        }

        private void SetImpl<T>(string Name, T Obj, bool IsGeneric)
        {
            var WasRunning = this.IsRunning;
            var IsRefType = !typeof(T).IsValueType & (typeof(T) != typeof(string));

            if (!WasRunning)
            {
                this.OnSetStarted();

                this.IsRunning = true;
                Assert.True(this.GetCache.Count == 0);
                Assert.True(this.SetCache.Count == 0);
                Assert.True(this.ObjectsCount == 0);
            }

            this.OnSetEnterContext(Name);

            try
            {
                var Id = this.ObjectsCount;
                if (IsRefType)
                {
                    var WasInCache = this.SetCache.TryGetValue(Obj, out var Tmp);

                    var IsDone = Tmp.Item2;
                    if (WasInCache)
                    {
                        Id = Tmp.Item1;
                    }

                    // We allow more than one serialization iterations on the same object. See the serializer for Object for a use case.

                    this.Set(nameof(Id), Id);

                    if (IsDone)
                    {
                        return;
                    }
                    else if (!WasInCache)
                    {
                        this.SetCache.Add(Obj, (Id, false));
                        this.ObjectsCount += 1;
                    }
                }

                var S = this.GetSerializer<T>(Obj, IsGeneric);
                if (IsGeneric)
                {
                    if (S is Serializer<T> ST)
                    {
                        ST.SetT(this.SetProxy, Obj);
                    }
                    else
                    {
                        S.Set(this.SetProxy, Obj);
                    }
                }
                else
                {
                    this.Set(nameof(Serializer), S);
                    S.Set(this.SetProxy, Obj);
                }

                if (IsRefType)
                {
                    Assert.True(this.SetCache.ContainsKey(Obj));
                    this.SetCache[Obj] = (Id, true);
                }
            }
            finally
            {
                this.OnSetExitContext(Name);

                if (!WasRunning)
                {
                    this.OnSetFinished();

                    this.IsRunning = false;
                    this.SetCache.Clear();
                    this.ObjectsCount = 0;
                }
            }
        }

        private T GetImpl<T>(string Name, CNullable<T> GObj, bool IsGeneric)
        {
            var WasRunning = this.IsRunning;
            var IsRefType = !typeof(T).IsValueType;

            // We allow T to be a value type. See the serializer for SerializationArrayChunk for a use case.

            if (!WasRunning)
            {
                this.OnGetStarted();

                this.IsRunning = true;
                Assert.True(this.GetCache.Count == 0);
                Assert.True(this.SetCache.Count == 0);
                Assert.True(this.ObjectsCount == 0);
            }

            this.OnGetEnterContext(Name);

            try
            {
                var Id = 0;
                if (IsRefType)
                {
                    Id = this.Get<int>(nameof(Id));

                    if (this.GetCache.TryGetValue(Id, out var Obj) && Obj != null)
                    {
                        return (T) Obj;
                    }

                    this.GetCache.Add(Id, null);
                }

                var S = default(Serializer);
                if (IsGeneric)
                {
                    S = this.GetSerializer<T>(default, true);
                }
                else
                {
                    S = this.Get<Serializer>(nameof(Serializer));
                }

                var ST = S as Serializer<T>;
                var R = default(T);

                if (IsGeneric & ST != null)
                {
                    if (GObj.HasValue)
                    {
                        ST.GetT(this.GetProxy, GObj.Value);
                        R = GObj.Value;
                    }
                    else
                    {
                        R = ST.GetT(this.GetProxy);
                    }
                }
                else if (GObj.HasValue)
                {
                    S.Get(this.GetProxy, GObj.Value);
                    R = GObj.Value;
                }
                else
                {
                    R = (T) S.Get(this.GetProxy);
                }

                if (IsRefType)
                {
                    var Obj = this.GetCache[Id];
                    if (Obj != null)
                    {
                        Verify.True(Obj == (object) R, "Two different objects returned for the same id.");
                    }
                    else
                    {
                        this.GetCache[Id] = R;
                    }
                }

                return R;
            }
            finally
            {
                this.OnGetExitContext(Name);

                if (!WasRunning)
                {
                    this.OnGetFinished();

                    this.IsRunning = false;
                    this.GetCache.Clear();
                    this.ObjectsCount = 0;
                }
            }
        }

        protected internal void Set<T>(string Name, T Obj)
        {
            this.SetImpl(Name, Obj, true);
        }

        protected internal void Set(string Name, object Obj)
        {
            this.SetImpl(Name, Obj, false);
        }

        protected internal T Get<T>(string Name)
        {
            return this.GetImpl<T>(Name, default, true);
        }

        protected internal object Get(string Name)
        {
            return this.GetImpl<object>(Name, default, false);
        }

        protected internal void Get<T>(string Name, T Obj)
        {
            this.GetImpl<T>(Name, Obj, true);
        }

        protected internal void Get(string Name, object Obj)
        {
            this.GetImpl(Name, new CNullable<object>(Obj), false);
        }

        public SerializerCollection Serializers { get; } = new SerializerCollection();

        private readonly FormatterSetProxy SetProxy;
        private readonly FormatterGetProxy GetProxy;

        private readonly Dictionary<object, (int, bool)> SetCache = new Dictionary<object, (int, bool)>(ReferenceEqualityComparer<object>.Instance);
        private readonly Dictionary<int, object> GetCache = new Dictionary<int, object>();
        private int ObjectsCount;
        private bool IsRunning;
    }
}
